### ISSUE TITLE
CMSD beaconing support

### ISSUE THREAD

**Issue by nicoweilelemental** [cta-wave/common-media-client-data#155]
CMSD implementations start to appear on the origin side: USP has got an experimental project ([github](https://github.com/unifiedstreaming/origin-cmsd)), and AWS Elemental MediaPackage has released CMSD support in GA in November 2024, with both [standard keys](https://docs.aws.amazon.com/mediapackage/latest/userguide/cmsd.html) and [custom keys](https://docs.aws.amazon.com/mediapackage/latest/userguide/mqcs.html) for media quality confidence score. 

The player will receive the CMSD keys through CMSD-Static and CMSD-Dynamic headers, and this information will be useful to 3rd party QoE analysis and content steering solutions. It is therefore important that CMCD beaconing can carry these two CSMD headers payload - in both CMCDv2 Request Mode and CMCDv2 Response Mode. It doesn't make sense in CNCDv2 State-Interval Mode, as the CMSD data is related to specific segments.

I'm suggesting to create two new CMCDv2 keys of type String, which value containing the base64-encoded value of the respective CSMD header:
- cmsds: to carry the CMSD-Static header(s) value
- cmsdd: to carry the CMSD-Dynamic header(s) value

It is possible that multiple CMSD header of the same type will be attached to a given object delivered to the player (especially the CMSD-Dynamic header). In this case, the player should concatenate the value of all CMSD headers of the same type, add a comma between these multiple values, and base64-encode the whole string. The header names ('CMSD-Static' or 'CMSD-Dynamic') should not be part of the CMCD cmsds and cmsdd keys value, only their values.

**Example of CMSD header**:
cmsd-static: n="MediaPackage:us-east-1:reference-streams-v2:CMAF-Ingest-Channel:CMAF-endpoint",st=l,sf=(d h),d=6000,br=5000000,ot=v,at=1733114706553,com.amazonaws-mqcs="100",com.amazonaws-mqcs-seq="100",com.amazonaws-mqcs-seq-tracks="v:100;a:100"

**How it would translate in CMCDv2**:
CMCD-Object:cmsds="bj0iTWVkaWFQYWNrYWdlOnVzLWVhc3QtMTpyZWZlcmVuY2Utc3RyZWFtcy12MjpDTUFGLUluZ2VzdC1DaGFubmVsOkNNQUYtZW5kcG9pbnQiLHN0PWwsc2Y9KGQgaCksZD02MDAwLGJyPTUwMDAwMDAsb3Q9dixhdD0xNzMzMTE0NzA2NTUzLGNvbS5hbWF6b25hd3MtbXFjcz0iMTAwIixjb20uYW1hem9uYXdzLW1xY3Mtc2VxPSIxMDAiLGNvbS5hbWF6b25hd3MtbXFjcy1zZXEtdHJhY2tzPSJ2OjEwMDthOjEwMCI="

**Comment by gwendalsimon** [comment]
Good idea. Note that, in the paper "[Proposal for CMSD-based transmission of Media Quality Assessment (MQA) data](https://docs.google.com/document/d/1ncNNitIY-1mpVoWASM_BVe8LL1WUAGU0lUqN2latKpw/edit?usp=sharing)" from @wilaw, one paragraph says: "_An alternate workflow is for the origin to still publish the MQA score in a CMSD response header, but it is ignored by the CDN and passed through to the client. The end client can then retrieve the MQA data and either send it to a 3rd party analytics solution, or pass it back to the CDN as Common Media Client Data (CMCD) fields_". This implementation is our favorite choice at Synamedia.

Though, a point of worries on the global size of CMCDv2. We are adding a lot of information in the CMCD reports. Here, an Origin can add many proprietary key/val (as in the example), which the client would necessarily report back.

**Comment by nicoweilelemental** [comment]
Good feedbacks.
@gwendalsimon I think your concern could be addressed if we had an additional boolean key to explicitly activate CMSD beaconing. It is true that not every implementer will systematically need these CMSD payloads or all use cases. 

@wilaw to your points:
1. Fair enough, there should be no key conflicts between all CMSD headers
2. base64 encoding choice is to prevent potential parsing problems in the chain. It's also potentially more compact than URL encoding.
3. With the additional key mentioned above we could toggle the beaconing on/off as needed. CMSD keys selection could be another way to reduce the volume if not all keys are of interest for the beaconing target system (see point 5 below)
4. Good point. The player should add a url key in the CMSD payload before the beaconing (we should also make it an optional key  in the CMSD spec)
5. I can think of two options here: either the player has got a configuration for the list of CMSD keys that needs to be beaconed, or we add a new CMCD key (like cmsdkl for 'CMSD keys list') that actually contains the list of CMSD keys that we want the player to add in the beaconed payload.

**Comment by PCaponetti** [comment]
Paul C: the whole payload is too large, what are the use cases?
Will: media quality score
Paul C: I also see upstream under distress and other keys that would indicate reasons for why the QoE is the way it is
Will: can configure which CMSD keys to echo addressing size
Paul C/Chris L: we've resisted connecting CMCD and CMSD in the past, may be due to actuations or they were resisted in v1 because the CDN will have just sent CMSD and should already know it. Those motivations may no longer apply.
Chris L: URI encoded would be smaller but we would likely need base64 encoded.
Nicolas L: Should specify that it's only for response mode, not request or event mode. Also we can just lift and shift if we add a cmcd key for each cmsd key.
Chris L: Think that is too complicated, would just drop the entire static in a static key and entire dynamic in a dynamic key.

Added keys:

CMSD Dynamic Header | cmsdd | CMCD-Request | String | Holds a Base64 encoded copy of the CMSD data received on the CMSD-Dynamic response header. This key MUST only be used in RESPONSE mode. 

CMSD Static Header | cmsds | CMCD-Request | String | Holds a Base64 encoded copy of the CMSD data received on the CMSD-Static response header. This key MUST only be used in RESPONSE mode. | No
